lms/djangoapps/courseware/views/views.py

```diff
@@ -749,40 +743,121 @@ def _progress(request, course_key, student_id):
+
+    if enrollment_mode == CourseMode.AUDIT:
+        return CertData(
+            CertificateStatuses.audit_passing,
+            'Your enrollment: Audit track',
+            'You are enrolled in the audit track for this course. The audit track does not include a certificate.',
+            download_url=None,
+            show_cert_web_view=False,
+            cert_web_view_url=None
+        )
+
+    if not show_generate_cert_btn:
+        return None
+
+    if certs_api.is_certificate_invalid(student, course_key):
+        return CertData(
+            CertificateStatuses.invalidated,
+            'Your certificate has been invalidated',
+            'Please contact your course team if you have any questions.',
+            download_url=None,
+            show_cert_web_view=False,
+            cert_web_view_url=None
+        )
+
+    cert_downloadable_status = certs_api.certificate_downloadable_status(student, course_key)
+    # If the learner is in verified modes and the student did not have
+    # their ID verified, we need to show message to ask learner to verify their ID first
+    missing_required_verification = enrollment_mode in CourseMode.VERIFIED_MODES and \
+        not SoftwareSecurePhotoVerification.user_is_verified(student)
+
+    if cert_downloadable_status['is_downloadable']:
+        cert_status = CertificateStatuses.downloadable
+        title = 'Your certificate is available'
+        msg = 'You can keep working for a higher grade, or request your certificate now.'
+        download_url = cert_downloadable_status['download_url']
         if certs_api.has_html_certificates_enabled(course_key, course):
             if certs_api.get_active_web_certificate(course) is not None:
-                context.update({
-                    'show_cert_web_view': True,
-                    'cert_web_view_url': certs_api.get_certificate_url(course_id=course_key, uuid=cert_status['uuid']),
-                })
+                cert_web_view_url = certs_api.get_certificate_url(
+                    course_id=course_key, uuid=cert_downloadable_status['uuid']
+                )
+                return CertData(
+                    cert_status, title, msg,
+                    download_url=None, show_cert_web_view=False, cert_web_view_url=cert_web_view_url
+                )
             else:
-                context.update({
-                    'is_downloadable': False,
-                    'is_generating': True,
-                    'download_url': None
-                })
+                return CertData(
+                    CertificateStatuses.generating,
+                    "We're working on it...",
+                    "We're creating your certificate. You can keep working in your courses and a link "
+                    "to it will appear here and on your Dashboard when it is ready.",
+                    download_url=None,
+                    show_cert_web_view=False,
+                    cert_web_view_url=None
+                )
```

Comment:
It will fall through the `return CertData` with `download_url` in the case when `certs_api.has_html_certificates_enabled(course_key, course)` returns False.